src/Datadog.Trace.ClrProfiler.Managed/Integrations/AspNet/AspNetCoreMvc2Integration.cs

```
@@ -0,0 +1,324 @@
+            {
+                // integration disabled
+                return;
+            }
+
+            string methodDef = $"{DiagnosticSource}.{nameof(BeforeAction)}(...)";
+            var integrationContext = AspNetAmbientContext.RetrieveFromHttpContext(httpContext);
+
+            if (integrationContext == null)
+            {
+                Log.Error($"Could not access {nameof(AspNetAmbientContext)} for {methodDef}.");
+            }
+            else
+            {
+                SetAspNetCoreMvcSpecificData(
+                    ambientContext: integrationContext,
+                    actionDescriptor: actionDescriptor,
+                    httpContext: httpContext);
+            }
+
+            Action<object, object, object, object> instrumentedMethod = null;
+
+            try
+            {
+                instrumentedMethod =
+                    MethodBuilder<Action<object, object, object, object>>
+                       .Start(Assembly.GetCallingAssembly(), mdToken, opCode)
+                       .WithConcreteTypeName(DiagnosticSource)
+                       .WithParameters(diagnosticSource, actionDescriptor, httpContext, routeData)
+                       .Build();
+            }
+            catch (Exception ex)
+            {
+                // profiled app will continue working as expected without this method
+                Log.ErrorException($"Error resolving {DiagnosticSource}.{nameof(BeforeAction)}(...)", ex);
+            }
+
+            try
+            {
+                // call the original method, catching and rethrowing any unhandled exceptions
+                instrumentedMethod?.Invoke(diagnosticSource, actionDescriptor, httpContext, routeData);
+            }
+            catch (Exception ex)
+            {
+                // profiled app will continue working as expected without this method
+                Log.Error($"Exception when calling {methodDef}.", ex);
```

Should we rethrow the original exception here?

---

No, this is a carry forward from the original implementation of this integration.
This is a diagnostic method which does not affect the behavior of user code.
I am okay with throwing, but it seems safer to just log.

---

We shouldn't change application behavior. If the instrumented method throws an exception, we should let it bubble up as if tracing was not enabled.